src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToLinqMethodProvider.cs

```diff
@@ -0,0 +1,169 @@
+                    case SyntaxKind.GroupClause: expression = ProcessGroupClause(expression, context.Identifier, (GroupClauseSyntax)selectOrGroupClause); break;
+                    case SyntaxKind.LetClause: return null; // Skip this because the linq method for the let will be more complicated than the query.
+                    default: return null;
+                }
+
+                if (queryBody.Continuation != null)
+                {
+                    context = new ConversionContext(queryBody.Continuation.Identifier, expression);
+                    return ProcessQueryBody(context, queryBody.Continuation.Body);
+                }
+
+                return expression;
+            }
+
+            private static ExpressionSyntax ProcessGroupClause(ExpressionSyntax expression, SyntaxToken identifier, GroupClauseSyntax groupClause)
+            {
+                var parameter = SyntaxFactory.Parameter(identifier);
+                var groupExpressionLambda = SyntaxFactory.SimpleLambdaExpression(parameter, groupClause.GroupExpression.WithoutTrailingTrivia());
+                var byExpressionLambda = SyntaxFactory.SimpleLambdaExpression(parameter, groupClause.ByExpression.WithoutTrailingTrivia());
+                var groupExpressionArgument = SyntaxFactory.Argument(groupExpressionLambda);
+                var byExpressionArgument = SyntaxFactory.Argument(byExpressionLambda);
+                var arguments = new SeparatedSyntaxList<ArgumentSyntax>().Add(byExpressionArgument).Add(groupExpressionArgument);
+
+                return CreateInvocationExpression(expression, arguments, "GroupBy");
+            }
+
+            private static ExpressionSyntax ProcesSelectClause(ExpressionSyntax expression, SyntaxToken identifier, SelectClauseSyntax selectClause)
+            {
+                var selectExpression = selectClause.Expression;
+                // Avoid trivial Select(x => x)
+                if (selectExpression is IdentifierNameSyntax identifierName)
+                {
+                    // TODO consider a better condition to compare
+                    if (identifierName.Identifier.Text == identifier.Text)
+                    {
+                        return expression;
+                    }
+                }
+
+                return CreateInvocationExpression(expression, identifier, selectClause.Expression, "Select");
+            }
+
+            private static ExpressionSyntax CreateInvocationExpression(ExpressionSyntax expression, SyntaxToken identifier, ExpressionSyntax bodyExpression, string keyword)
+            {
+                var parameter = SyntaxFactory.Parameter(identifier);
+                var lambda = SyntaxFactory.SimpleLambdaExpression(parameter, bodyExpression.WithoutTrailingTrivia());
+                var argument = SyntaxFactory.Argument(lambda);
+                var arguments = new SeparatedSyntaxList<ArgumentSyntax>().Add(argument);
```

Comment:
use SyntaxFactory.SeparatedList(argument)

Reply:
Consider just making this into one large expression.

Reply:
would use syntax factory but still have to use Add

---
In reply to: [167387622](https://github.com/dotnet/roslyn/pull/24759#discussion_r167387622) [](ancestors = 167387622)

Reply:
what is the benefit of doing this?

---
In reply to: [167387632](https://github.com/dotnet/roslyn/pull/24759#discussion_r167387632) [](ancestors = 167387632)

Reply:
> what is the benefit of doing this?

Separate statements makes it less clear (to me) what's going on.  You have to analyze each variable to see how it is used.  Merging into a single expression makes it clear how the construction happens and conveys the structural nature of the tree you're creating.  This is what we do elsewhere throughout our codegen features.

Reply:
OK. Agree. Thank you! Made some squeezing here. Please let me know if you see more reasonable ones.

---
In reply to: [167747706](https://github.com/dotnet/roslyn/pull/24759#discussion_r167747706) [](ancestors = 167747706)